### ISSUE TITLE
CrmServiceClient ImportSolutionToCrm Hangs on Some Machines

### ISSUE THREAD

**Issue by programmergrrl** [seanmcne/Microsoft.Xrm.Data.PowerShell#342]
I don't really think this is an issue with Microsoft.Xrm.Data.PowerShell, as I can reproduce it using just CrmServiceClient's ImportSolutionToCrm method directly, but I'm curious if it sounds familiar to you, or if you have any ideas about how to debug it further? The import method hangs when I try to run it on some machines, but not others. 

I have it written up over here on the Dynamics 365 forum: https://community.dynamics.com/365/f/761/t/301382

**Comment by seanmcne** [comment]
Can you try importing using the asybc option and see if that is success?

Also, does this repro for all solutions or just larger ones?

**Comment by programmergrrl** [comment]
Yes, we tried the async option, and that is a success, except we are facing a separate issue with it, where in 3 Dynamics 365 instances, it encounters a SqlTimeout exception. I believe this is a separate issue, because for those same instances, if I use the synchronous method from my developer machine, the import works just fine. So, I have a separate ticket open with Microsoft support to investigate the SqlTimeout on those instances.

Silly question, in your experience, for a large solution, it is better to use the synchronous method or the async job, or doesn't it matter? When I google it, I find articles from around the time the async was added saying it's a good idea and _less_ likely to encounter SqlTimeout exceptions. But, I also notice in the release notes for Microsoft.CrmSdk.XrmTooling.CoreAssembly v9.0.2.4, in the notes about ImportSolutionToCrmAsync, it says "Note, Async jobs are subject to Async queueing and overhead, thus use this with caution," so I'm a bit confused.

With regard to my original issue about the import hanging, it only happens for our large solution. When I try it with a small test solution, it works fine on all machines.

One thing I noticed that may work for us is to use a slight modification to the Microsoft.Xrm.Data.PowerShell Import-CrmSolution code. If I set the timeout on the CrmServiceClient to be something reasonably short, say 5 minutes, so that it times out after the solution has uploaded but while it is still importing, then if I just change the line checking the result of the solution import from this
```
if ($result -eq [guid]::Empty) {
        throw LastCrmConnectorException($conn)
}
```
to this
```
if ($result -eq [guid]::Empty -and $importId -eq [guid]::Empty) {
        throw LastCrmConnectorException($conn)
}
```
then everything works. ImportSolutionToCrm has set the import ID and the powershell code then uses it to monitor it to completion.

So, I'm curious about the powershell code that calls ImportSolutionToCrm, because as I've been debugging this, in our environments, that call doesn't appear to return until the solution has finished importing, so there would be no need to subsequently poll for the import to complete. Is our environment behaving incorrectly? Should it be returning once the file has uploaded but before the import has completed? Or is that logic for a different version of Dynamics 365? (We're using v9 online.)

**Comment by seanmcne** [comment]
While I've seen imports seem to hang I haven't been able to reproduce it easily but I have been able to always successfully import using the async option.  To help maintain the concept of uploading and "waiting" I've added a timeout where it will poll the job uploaded to the system and report the progress being made server-side rather than holding a connection open to the server waiting on a response from a potential very long running operation.  I've observed the solution import issue but since I haven't been able to reproduce it at-will I haven't been able to debug it and only been able to nail it down to exceeding the network connection timeout and the CrmServiceClient not handling it well.  That said, it doesn't mean there isn't a bug or another scenario that would cause it to get hung up. 

One note, the last time I checked our max connection time in Online was 5 minutes (anything above would be forcibly closed by our network) - I suspect this may be behind the problems with Service Client being able to do large solution imports without asynchronously executing them. 

As for your final note about the Import-SolutionToCrm code - when calling $conn.ImportSolutionToCrm() - that is a request/response type of api call where the result of that will either be a success or a failure - it doesn't return anything back until it has completed. 

What is a little confusing is that I've added a switch to use sync for this function by default, but you can alternatively redirect it (using the same params) to execute the async function - this was done to try and make it easy for folks to quickly try async vs. sync imports without having to churn their scripts too much but the trade off is confusing. 

To add a little more confusion, unfortunately, even after a synchronous import has completed, sometimes there are some follow up steps that don't always seem to be completed so there's a subsequent polling that occurs to ensure we see success or failure (it's a terrible terrible pattern but CrmServiceClient doesn't give us much of an option on this and it was the only thing I could think of to give a higher quality result at the time).  We've also seen the sync import return an import Id yet the import actually failed or is still partially completed: this is why the code goes back to read the results of the manifest.  That was done a while back so it's possible this was due to a product 'quirk' or bug but I found getting an id or not wasn't a valid test for a solution import success or failure. 

That said, I'm not saying it can't be improved :) - I like the idea of checking for the importid being empty as well as that might indicate it truly failed (although I've seen both be blank, yet succeed later on in the manifest).  It might be an acceptable tradeoff though to assume if both are null or empty then we should just assume it's a total failure.

**Comment by seanmcne** [comment]
PS: one thing I forgot to mention, the async import method is usually better to execute directly than to use the switch parameter.  The primary advantage to using Import-CrmSolutionAsync is that it will report back verbose status messages so you can see what it's doing and what's going on if it takes an excessive amount of time where the sync import with the async switch will dump the results out after it runs.  

![image](https://user-images.githubusercontent.com/2292260/48634419-1b25fb00-e98b-11e8-889f-64a41147870a.png)

**Comment by programmergrrl** [comment]
Thanks for your help. I wish I could shed more light on how to reproduce the hanging import, but it's one of those situations where it works on one machine and not another and I don't know enough about it to say why. 

No worries about using the async import method vs the switch--we already saw that's the preferred way and used it when we tried it async.

I think we're going ahead with checking for the importid being empty as well, then, but I'm not sure it's something everyone would want, just something that seems to be working in our current situation. I'm glad you've already got the code in there to do all the checking for the import progress.

Thanks again :-)
